From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Mon, 29 Sep 2025 19:47:55 +0200
Subject: wayland: Always send configure event after xdg_popup::reposition

Commit 6c7565eee5aa6f1cee33c11c100f806aa421b619 changes
meta_window_wayland_move_resize_internal so that it doesn't send a configure
event if there were no changes since the last sent configure event.

However the Wayland spec mandates that xdg_popup::repositioned and
xdg_popup.configure be sent after a xdg_popup::reposition request, even if
there were no changes in the placement.

Add a new META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE flag to indicate that we
require sending a configure event no matter what.

This fixes a bug where entering a submenu within Gtk.PopoverMenu hangs GTK
waiting for a configure event that never comes.

Bug-Ubuntu: https://launchpad.net/bugs/2125770
Origin: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4691
---
 src/core/window-private.h         |  1 +
 src/wayland/meta-window-wayland.c | 13 +++++++++----
 2 files changed, 10 insertions(+), 4 deletions(-)

diff --git a/src/core/window-private.h b/src/core/window-private.h
index db69a06..42760bd 100644
--- a/src/core/window-private.h
+++ b/src/core/window-private.h
@@ -74,6 +74,7 @@ typedef enum
   META_MOVE_RESIZE_WAYLAND_CLIENT_RESIZE = 1 << 12,
   META_MOVE_RESIZE_CONSTRAIN = 1 << 13,
   META_MOVE_RESIZE_RECT_INVALID = 1 << 14,
+  META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE = 1 << 15,
 } MetaMoveResizeFlags;
 
 typedef enum _MetaPlaceFlag
diff --git a/src/wayland/meta-window-wayland.c b/src/wayland/meta-window-wayland.c
index d6a58f7..9497fe3 100644
--- a/src/wayland/meta-window-wayland.c
+++ b/src/wayland/meta-window-wayland.c
@@ -330,7 +330,8 @@ should_configure (MetaWindow          *window,
 
   /* The state was changed, or the change was explicitly marked as a configure
    * request. */
-  if (flags & META_MOVE_RESIZE_STATE_CHANGED)
+  if (flags & META_MOVE_RESIZE_STATE_CHANGED ||
+      flags & META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE)
     return TRUE;
 
   return FALSE;
@@ -445,6 +446,7 @@ meta_window_wayland_move_resize_internal (MetaWindow                *window,
             case META_PLACEMENT_STATE_CONSTRAINED_PENDING:
               {
                 if (flags & META_MOVE_RESIZE_PLACEMENT_CHANGED ||
+                    flags & META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE ||
                     !last_sent_configuration ||
                     rel_x != last_sent_configuration->rel_x ||
                     rel_y != last_sent_configuration->rel_y ||
@@ -460,7 +462,8 @@ meta_window_wayland_move_resize_internal (MetaWindow                *window,
                                                                       configured_rect.width,
                                                                       configured_rect.height,
                                                                       geometry_scale);
-                    if (!meta_wayland_window_configuration_is_equivalent (
+                    if (flags & META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE ||
+                        !meta_wayland_window_configuration_is_equivalent (
                           configuration,
                           wl_window->last_sent_configuration))
                       {
@@ -508,7 +511,8 @@ meta_window_wayland_move_resize_internal (MetaWindow                *window,
                                                    geometry_scale,
                                                    flags,
                                                    gravity);
-          if (!meta_wayland_window_configuration_is_equivalent (
+          if (flags & META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE ||
+              !meta_wayland_window_configuration_is_equivalent (
                 configuration,
                 wl_window->last_sent_configuration))
             {
@@ -1547,7 +1551,8 @@ meta_window_place_with_placement_rule (MetaWindow        *window,
     place_flags |= META_PLACE_FLAG_CALCULATE;
 
   meta_window_move_resize_internal (window,
-                                    (META_MOVE_RESIZE_MOVE_ACTION |
+                                    (META_MOVE_RESIZE_WAYLAND_FORCE_CONFIGURE |
+                                     META_MOVE_RESIZE_MOVE_ACTION |
                                      META_MOVE_RESIZE_RESIZE_ACTION |
                                      META_MOVE_RESIZE_PLACEMENT_CHANGED |
                                      META_MOVE_RESIZE_CONSTRAIN),
